traitsui/editor.py

```diff
@@ -377,32 +392,86 @@ def get_undo_item(self, object, name, old_value, new_value):
+        name : str
+            The extended name to parse.
+
+        Returns
+        -------
+        object, name, getter : any, str, callable
+            The object from the context, the (extended) name of the
+            attributes holding the value, and a callable which gets the
+            current value from the context.
         """
-        col = name.find(".")
-        if col < 0:
+        parts = name.split(".", 1)
+        if len(parts) == 1:
             object = self.context_object
         else:
-            object, name = self.ui.context[name[:col]], name[col + 1 :]
+            object_name, name = parts
+            object = self.ui.context[object_name]
 
-        return (object, name, eval("lambda obj=object: obj." + name))
+        return (object, name, partial(xgetattr, object, name))
 
     def _sync_values(self):
-        """ Initializes and synchronizes (as needed) editor traits with the
-            value of corresponding factory traits.
+        """ Initialize and synchronize editor and factory traits
+
+        Initializes and synchronizes (as needed) editor traits with the
+        value of corresponding factory traits.  The name of the factory
+        trait and the editor trait must match and the factory trait needs
+        to have ``sync_value`` metadata set.  The strategy followed is:
+
+        - for each factory trait with ``sync_value`` metadata:
+
+          1.  if the value is a :class:`ContextValue` instance then
+              call :meth:`sync_value` with the ``name`` from the
+              context value.
+          2.  if the trait has ``sync_name`` metadata, look at the
+              referenced trait value and if it is a non-empty string
+              then use this value as the name of the value in the
+              context.
+          3.  otherwise initialize the current value of the factory
+              trait to the corresponding value of the editor.
+
+        - synchronization mode in cases 1 and 2 is taken from the
+          ``sync_value`` metadata of the editor trait first and then
+          the ``sync_value`` metadata of the factory trait if that is
+          empty.
+
+        - if the value is a container type, then the `is_list` metadata
+          is set to
         """
         factory = self.factory
         for name, trait in factory.traits(sync_value=not_none).items():
             value = getattr(factory, name)
+            self_trait = self.trait(name)
+            if self_trait.sync_value:
+                mode = self_trait.sync_value
+            else:
+                mode = trait.sync_value
             if isinstance(value, ContextValue):
-                self_trait = self.trait(name)
                 self.sync_value(
                     value.name,
                     name,
-                    self_trait.sync_value or trait.sync_value,
-                    self_trait.is_list is True,
+                    mode,
+                    bool(self_trait.is_list),
```

Comment:
Is the `bool` call necessary? Does the `sync_value` method expect a strict boolean?

Reply:
Not necessary, but I think the value returned by `is_list` will frequently be `None` due to the way that trait metadata works, and given that the line was previously `self_trait.is_list is True` this is a more modern and safer rendering of the original code.

It's likely to save debugging time if someone runs into a problem with the code.